### PR TITLE
Fix `rustdoc` argument error

### PR DESCRIPTION
**Pull request by inashivb**
This is the fix is corresponding to https://github.com/rust-lang/rust/issues/88756 mentored by @jyn514

### CODE REVIEW

File: src/librustdoc/config.rs [hunk 1]
```diff
@@ -417,7 +417,12 @@ impl Options {
             }
             return Err(0);
         }
+        let (lint_opts, describe_lints, lint_cap) = get_cmd_lint_options(matches, error_format);
 
+        if lint_opts.is_empty() {
+            crate::usage("rustdoc");
+            return Err(0);
+        }
```
Comment (jyn514) [inline src/librustdoc/config.rs:425]: What does this do? Can you add a test showing the behavior?

**Review by jyn514 — COMMENTED** [review]
Please add a test for this - it would go somewhere in `run-make`, there should be existing rustdoc tests there you can use as examples.

**Comment by GuillaumeGomez** [comment]
Can you add a test in `src/test/rustdoc-ui` where you give no arguments to rustdoc please?

File: src/librustdoc/config.rs (hunk 1 above)
Comment (inashivb) [reply src/librustdoc/config.rs:425]: Hi @jyn514 !
I used this because `rustdoc` without any arguments and `rustdoc` with ambiguous arguments to an argument e.g. `rustdoc -W isrgnlk` had `matches.free` as an empty list. I looked for other params to differentiate between these two cases so that we print the default help in case no args were passed and error out in the latter case. I found that `lint_opts` was empty in case of no args and it was not empty when there was garbled output.

Do you think this is the right behavior for the two cases?

Note: Thanks a lot for the prompt review! :) I'm looking at how to add tests and keeping a check on any other feedback that you or GuillaumeGomez might have.

**Comment by inashivb** [comment]
> Can you add a test in `src/test/rustdoc-ui` where you give no arguments to rustdoc please?

Thank you for the prompt review! :)
I'm observing many existing tests failing on `x86_64-gnu-llvm-10, ubuntu-latest-xl` which were not really failing on my system, could you please tell me how should I investigate these?

**Comment by GuillaumeGomez** [comment]
Just in case you don't know how to run tests:

```console
$ x.py test src/test/rustdoc --stage 1
$ x.py test src/test/rustdoc-ui --stage 1
```

If one of the two doesn't work, it's very likely your change which broke them.

**Comment by inashivb** [comment]
> Just in case you don't know how to run tests:
> 
> ```
> $ x.py test src/test/rustdoc --stage 1
> $ x.py test src/test/rustdoc-ui --stage 1
> ```
> 
> If one of the two doesn't work, it's very likely your change which broke them.

Thank you. I indeed did not test the right way. I just ran `x.py test` and that seemed to work. I seem to have broken many things. I'll work on resolving them.

File: src/test/rustdoc-ui/issue-88756-default-output.rs [hunk 2]
```diff
@@ -0,0 +1,2 @@
+// compile-flags: -h
```
Comment (inashivb) [inline src/test/rustdoc-ui/issue-88756-default-output.rs:1]: Hi @GuillaumeGomez !
I'm unable to figure out how to make the test for no f;gas. I'm not sure even if this `-h` is correct here as the key says `compile-flags` but it did give the same results as `rustdoc -h`. Could you please point me to any existing similar tests?

Comment (GuillaumeGomez) [reply src/test/rustdoc-ui/issue-88756-default-output.rs:1]: That is actually a good question since the current file name is passed as argument... I can't find documentation about this either.

@jyn514: Do you know if there is a way to prevent any argument to be passed?

Comment (jyn514) [reply src/test/rustdoc-ui/issue-88756-default-output.rs:1]: You may need a run-make test instead of a UI test.

Comment (inashivb) [reply src/test/rustdoc-ui/issue-88756-default-output.rs]: Hi @jyn514 and @GuillaumeGomez !
I added the run-make test, could you please check if I did things the right way?

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me, thanks! Wanna take a last look @jyn514 ?

File: src/test/run-make/issue-88756-default-output/Makefile [hunk 3]
```diff
@@ -0,0 +1,4 @@
+-include ../../run-make-fulldeps/tools.mk
+
+all:
+	$(RUSTDOC) 2>&1 | diff - output-default.stdout
```
Comment (jyn514) [inline src/test/run-make/issue-88756-default-output/Makefile:4]: This is not testing the right thing. The original issue was that -W help didn't show the lint help.

File: src/librustdoc/config.rs [hunk 4]
```diff
@@ -417,7 +417,12 @@ impl Options {
             }
             return Err(0);
         }
+        let (lint_opts, describe_lints, lint_cap) = get_cmd_lint_options(matches, error_format);
 
+        if matches.free.is_empty() && lint_opts.is_empty() {
+            crate::usage("rustdoc");
+            return Err(0);
```
Comment (jyn514) [inline src/librustdoc/config.rs:424]: Why was this change necessary? I'm pretty sure it had this behavior even before.

**Review by jyn514 — CHANGES_REQUESTED** [review]

File: src/librustdoc/config.rs (hunk 4 above)
Comment (inashivb) [reply src/librustdoc/config.rs:424]: Hi @jyn514 !
Apologies for late response, my dev env was broken.
If I just move the `get_cmd_lint_options`, I still get:
```
└─ $ ▶ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustdoc 
error: missing file operand
```
So, I checked for the `matches` attributes as the condition leading to this message was checking whether `matches.free` was empty. I found out that `matches.free` was indeed empty even after this move but what was also different with this situation compared to other places where `matches.free` is expected to error out was that `lint_opts` was also empty.

Before this change,
```
└─ $ ▶ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustdoc -W help
error: missing file operand
```
and 
```
└─ $ ▶ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustdoc -W
error: Argument to option 'W' missing
```

After adding these condition checks,
```
└─ $ ▶ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustdoc -W help
rustdoc [options] <input>

Options:
    -h, --help          show this help message
    -V, --version       print rustdoc's version
    -v, --verbose       use verbose output
    -r, --input-format [rust]
                        the input type of the specified file
    -w, --output-format [html]
                        the output type to write
    -o, --output PATH   where to place the output
        --crate-name NAME
                        specify the name of this crate
        --crate-type [bin|lib|rlib|dylib|cdylib|staticlib|proc-macro]
                        Comma separated list of types of crates
                        for the compiler to emit
    -L, --library-path DIR
                        directory to add to crate search path
        --cfg           pass a --cfg to rustc
        --extern NAME[=PATH]
```
and
```
└─ $ ▶ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustdoc -W
error: Argument to option 'W' missing
```
Is it still incorrect behavior? I'm not sure what changed from before if it was working with just the move.

File: src/test/run-make/issue-88756-default-output/Makefile (hunk 3 above)
Comment (inashivb) [reply src/test/run-make/issue-88756-default-output/Makefile:4]: OK. Shall I add test for default behavior and `-W help` as both seem to have been mentioned?

File: src/librustdoc/config.rs (hunk 4 above)
Comment (jyn514) [reply src/librustdoc/config.rs:424]: > Is it still incorrect behavior?

You can test by checking against rustc :) rustdoc and rustc should behave the same for invalid arguments.

(I haven't checked myself, I can do so if you have trouble.)

Comment (inashivb) [reply src/librustdoc/config.rs:424]: hmm I am a little confused now. It seems as if rustdoc already (i.e. without this patch) works like rustc.
For example:
```
└─ $ ▶ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustc -L help
error: no input filename given
```
It does not however behave this way particularly for 4 options (out of which one is `-W`) which are mentioned in the "additional help section"
```
Additional help:
    -C help             Print codegen options
    -W help             Print 'lint' options and default settings
    -Z help             Print unstable compiler options
    --help -v           Print the full set of options rustc accepts
```
So, if I were to pass `help` as mentioned in the cmd output above with either of these 4 options, I'll get some help corresponding to that option, e.g.
```
└─ $ ▶ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustc -W help

Available lint options:
    -W <foo>           Warn about <foo>
    -A <foo>           Allow <foo>
    -D <foo>           Deny <foo>
    -F <foo>           Forbid <foo> (deny <foo> and all attempts to override)


Lint checks provided by rustc:

                                                       name  default  meaning
                                                       ----  -------  -------
                     absolute-paths-not-starting-with-crate  allow    fully qualified paths that start with a module name instead of `crate`, `self`, or an extern crate name
                                               box-pointers  allow    use of owned (Box type) heap memory
                                       deprecated-in-future  allow    detects use of items that will be deprecated in a future version
                                  elided-lifetimes-in-paths  allow    hidden lifetime parameters in types are deprecated
```
But, it seems that rustdoc does not promise any additional help or has a separate document for any of its options. Also, `-W` seems to mean different things for rustc and rustdoc.

The only difference that I am able to see is that rustdoc by default also errors out
```
─ $ ▶ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustdoc 
error: missing file operand
```
while rustc prints help
```
└─ $ ▶ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustc
Usage: rustc [OPTIONS] INPUT

Options:
    -h, --help          Display this message
        --cfg SPEC      Configure the compilation environment
    -L [KIND=]PATH      Add a directory to the library search path. The
                        optional KIND can be one of dependency, crate, native,
                        framework, or all (the default).
    -l [KIND[:MODIFIERS]=]NAME[:RENAME]
                        Link the generated crate(s) to the specified native
                        library NAME. The optional KIND can be one of
                        static, framework, or dylib (the default).
                        Optional comma separated MODIFIERS
                        (bundle|verbatim|whole-archive|as-needed)
                        may be specified each with a prefix of either '+' to
                        enable or '-' to disable.
        --crate-type [bin|lib|rlib|dylib|cdylib|staticlib|proc-macro]
                        Comma separated list of types of crates
                        for the compiler to emit
        --crate-name NAME
```

Did I do what you asked me to check, @jyn514 ? If yes, where does it make us stand on this current rustdoc issue (problem with `-W help`)?

Comment (jyn514) [reply src/librustdoc/config.rs:424]: > Is it still incorrect behavior? I'm not sure what changed from before if it was working with just the move.

Yes, it's still correct - it should be showing the same lint help as rustc for `-W help`, not the generic usage.

> But, it seems that rustdoc does not promise any additional help or has a separate document for any of its options. Also, -W seems to mean different things for rustc and rustdoc.

That should not be the case. If rustdoc doesn't give help for `-W help`, that's also a bug that should be fixed.

> The only difference that I am able to see is that rustdoc by default also errors out while rustc prints help

We should fix that too.

> hmm I am a little confused now. It seems as if rustdoc already (i.e. without this patch) works like rustc (for the -W help like scenarios).

I'm not sure why you think `-L` and `-W` are similar. `rustc --help` says
```
    -L [KIND=]PATH      Add a directory to the library search path. The
                        optional KIND can be one of dependency, crate, native,
                        framework, or all (the default).
...

Additional help:
    -C help             Print codegen options
    -W help             Print 'lint' options and default settings
    -Z help             Print unstable compiler options
    --help -v           Print the full set of options rustc accepts
```
`-L help` is a valid option that means "add the directory named help/ to the search path".

**Comment by jyn514** [comment]
Hi @inashivb do you know how to make progress here? Do you need help?

**Comment by inashivb** [comment]
> Hi @inashivb do you know how to make progress here? Do you need help?

Hi @jyn514 ! Yes. Could you please tell where in rustc code can I find the cmdline arguments setting and rendering? I could not find any usage of  `crate::usage` or the help strings for any option.

**Comment by jyn514** [comment]
@inashivb I think you're looking for https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_driver/lib.rs.html#985

**Comment by inashivb** [comment]
@jyn514 apologies for the massive delay. Looking at it this weekend.

**Comment by JohnCSimon** [comment]
Ping from triage: 
@inashivb What is the status of this PR? 
.Please adjust the label with `@rustbot ready` when this is ready for review.

**Comment by inashivb** [comment]
@JohnCSimon it's not ready yet. Sorry about this. Looking at it today. Trying to match rustc's behavior here.

**Comment by jyn514** [comment]
FYI I've been chatting with @inashivb in DM's, they're still working on this :) no need to triage for now.

**Comment by jyn514** [comment]
I am no longer on the Rustdoc team. You can assign a new reviewer with

r? rust-lang/rustdoc

**Comment by notriddle** [comment]
It looks like some new arguments were added, and the new output needs blessed.

**Comment by Dylan-DPC** [comment]
@inashivb any updates on this?

**Comment by inashivb** [comment]
> @inashivb any updates on this?

Hi @Dylan-DPC !

I'm afraid I'm not sure how to deal with the issue that is there now. I'll try to explain:
In order to show the default help output in case no argument was given to `rustdoc` on cmdline, I check for the length of the arguments given on the cmdline (following how `rustc` does it), if it was 0, we output default help else let the existing code take care of whatever arg was passed.
However, when I run `x.py test --stage 1 tests/../mytest` I observe that `rustdoc` is supplied with some args like `-L` by default. I pass no arguments to the test (https://github.com/rust-lang/rust/pull/88831/commits/69e6d4a7e936744e31b14c0c0ca47f28cab22f5a#diff-0fa1bfe4e7dc2832e4d253c4bd571dc4a223b0d5f874e2bf4e3a5ed6889b5fc8R4) and it still has at least the library in the args so the check I have for `args.is_empty()` fails.

In conclusion,
If I check the generated `rustdoc` binary in `stage1` after building this code, it works just as expected. I give it no args and it prints the default help output. But, I cannot seem to convert the same behavior to test because of mandatory extra args passed during test.

Could you please tell what's the way to go about this?

**Comment by notriddle** [comment]
How about changing `$(RUSTDOC)` to `$(BARE_RUSTDOC)`?

**Comment by inashivb** [comment]
> How about changing `$(RUSTDOC)` to `$(BARE_RUSTDOC)`?

@notriddle thank you very much! That works locally at least. Here I see a bunch of more options that come with `rustc -W help` but not with `rustdoc -W help` when I run locally. Do you know what I might be doing wrng now?

**Review by notriddle — APPROVED** [review]
It looks fine to me, but I admit I’m not an expert on this part of the code.

**Comment by JohnCSimon** [comment]
triage:
@notriddle anyone else who can review this?

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me as well. Thanks @inashivb !

@bors: r=GuillaumeGomez,notriddle

**Comment by matthiaskrgr** [comment]
@bors r- 
looks like this ended up in a rollup  https://github.com/rust-lang/rust/pull/95288  while it failed already in its own ci

**Comment by Dylan-DPC** [comment]
@inashivb if you can fix the CI failures, we can get this merged again

**Comment by inashivb** [comment]
Hi @Dylan-DPC ! thanks for tagging me. I must have missed the notifications for change in labels. Could you please take a look at the branch now?

**Comment by inashivb** [comment]
I see the cause of failure as
```
< More information available at https://doc.rust-lang.org/1.62.0/rustdoc/what-is-rustdoc.html
---
> More information available at https://doc.rust-lang.org/nightly/rustdoc/what-is-rustdoc.html
```

Is there a VERSION or something env var? I tried looking for version in small and caps but can't see rust specific version setting. Could you please assist here?

**Comment by compiler-errors** [comment]
I did a [bors resync](https://github.com/rust-lang/rust/pull/94206#issuecomment-1120251623) and this seems to have gotten reapproved, sorry for the noise

@bors r-

**Comment by inashivb** [comment]
> Yeah, it's called `$CHANNEL`

@notriddle could you please assist in how to use it? I cannot seem to make it work by setting it on Makefile or the `main.rs` file (that I saw some UI tests do) [not sure if I'm reading it right though 😓 ]

**Comment by GuillaumeGomez** [comment]
I finished this PR in https://github.com/rust-lang/rust/pull/98331. 
Thanks @inashivb for working on this!
